### DISCUSSION TITLE
Next js 14.0.4  401 Unauthorised error when using fetch() in server component to get data from api route on same-origin

### DISCUSSION THREAD

**Discussion by ehnoeck** [vercel/next.js#59731] [Help]
### Summary

I have a server component making a request to api route on the same origin but on deployment to vercel it throws  401 Unauthorised error .
Suggestions for improvement are also much appreciated
"NB: the courses data in the api route are dummy and  are soon to be fetched from a db"
**THE API ROUTE**
```
import courses from "@/app/courses";
import { NextRequest, NextResponse } from "next/server";

export async function GET(request: NextRequest) {
  return NextResponse.json(courses);
}

```
**GETTING THE DATA**
```
async function getData(query: string) {
  let URL;
  if (process.env.environment === "dev") {
    URL = "http://" + process.env.URL + ":3000";
  } else {
    URL = "https://" + process.env.VERCEL_URL;
  }

  const url = `${URL}/explore/search`;

  try {
    const response = await fetch(url);

    if (!response.ok) {
      throw new Error(`HTTP error! Status: ${response.status}`);
    }
  } catch (error: any) {
    console.error("Fetch error:", error.message);
  }
  const response = await fetch(url);

  const results = await response.json();

  return results;
}
```

**Comment by gerkim62** [comment]
I have the same problem, didi you solve it

**Reply by ehnoeck** [reply]
I did not trace the actual cause of the unauthorised error but what I did to solve it was instead of fetching from a REST endpoint on the same app, just use server components to do the requests

**Reply by gerkim62** [reply]
After thinking about it for sometime, my problem was that I was using the VERCEL_URL (https://vercel.com/docs/projects/environment-variables/system-environment-variables) in my code, and since vercel generates multiple URLs for the same deployment, my server was trying to fetch from a different url than the one i am fetching although it is the same deployment. I fixed it by providing my own VERCEL_URL env variable instead of using the automatic one, i disabled "Automatically expose System Environment Variables" in my deployment settings. My project is now working in vercel. I hope this is helpful to anyone facing the same issue.

**Reply by ash-t-luu** [reply]
Hi all, I am having the same issue here. I tried following the creation of my env variable instead of using VERCEL's autogenerated one but am still running into the same issue.
@gerkim62 Did you set the value of your env var to the most recent deployment URL Vercel provided? 

@ehnoeck I am curious about what you mean by using server components to do the request. Do you have any examples I can refer to instead of using fetch()?

Thanks so much for your help.

**Reply by gerkim62** [reply]
I used the url that i am actually visiting instead of the the one that contains your github username. 

If it is okay for you to use server actions (ie the api will not be called from third party site or app) you may want to look at this video on getting started with server actions: https://www.youtube.com/watch?v=5MRLO-7O2So&ab_channel=BrettWestwood-SoftwareEngineer

in my case i have to call my api from my chrome extension that is why i needed it to be an api route. If for your case it does not need to be api route, the best way is to use server actions and server components.

**Reply by ash-t-luu** [reply]
Thank you @gerkim62. Will look into the video and more into this. Appreciate it.

**Reply by ehnoeck** [reply]
> Hi all, I am having the same issue here. I tried following the creation of my env variable instead of using VERCEL's autogenerated one but am still running into the same issue.
> @gerkim62 Did you set the value of your env var to the most recent deployment URL Vercel provided? 
> 
> @ehnoeck I am curious about what you mean by using server components to do the request. Do you have any examples I can refer to instead of using fetch()?
> 
> Thanks so much for your help.
I meant if you are making fetch calls to get data you could do it in the server component rather than calling from a route handler .
Watch the first segment of this vercel video
[]()https://youtu.be/RBM03RihZVs?si=Nm_An3TEPiVS7IlZ
